File: 2023/perl/day04a/day04a.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub usage {
    print "usage: $0 <file>\n";
    exit 1;
}

sub process {
    my ($contents) = @_;
    my $total = 0;
    my @lines = split(/\n/, $contents);
    foreach my $line (@lines) {
        my (undef, $rest) = split(':', $line, 2);
        my ($winning_str, $hand_str) = split(/\|/, $rest, 2);
        my %winning = map { $_ => 1 } split(' ', $winning_str);
        my %hand = map { $_ => 1 } split(' ', $hand_str);
        my $count = scalar(grep { exists $winning{$_} } keys %hand);
        $total += 2**($count - 1) if $count > 0;
    }
    return $total;
}

sub main {
    if (@ARGV < 1) {
        usage();
    }

    my $filename = $ARGV[0];

    open my $infile, '<', $filename or die "read of input file '$filename' failed.\n";
    my $contents = do { local $/; <$infile> };
    close $infile;

    my $result = process($contents);
    print "result = $result\n";
}

if ($0 eq __FILE__) {
    main();
}

File: 2023/ruby/day04a/day04a.rb
#!/usr/bin/env ruby

require 'set'

def usage
    progname = $PROGRAM_NAME
    puts "usage: #{progname} <file>"
    exit(1)
  end
  
def process(contents)
    result = 0
    contents.each_line do |line|
        rest = line.split(%r{:\s+})[1]
        winning_str = rest.split(%r{\s+\|\s+})[0]
        hand_str = rest.split(%r{\s+\|\s+})[1]
        winning_set = Set.new(winning_str.split().map(&:to_i))
        hand_set = Set.new(hand_str.split().map(&:to_i))
        count = (winning_set & hand_set).length
        if count > 0
            result += 2 ** (count - 1)
        end
    end
    result
  end
    
 def main()
    if ARGV.length < 1
      usage
    end
    
    filename = ARGV[0]
    contents = File.read(filename)
    result = process(contents)
    puts "result = #{result}"
  end
  
main()
